File: 151004/Tishalovich/rest_basics/publisher/src/main/java/by/bsuir/dc/publisher/services/impl/MessageServiceImpl.java
package by.bsuir.dc.publisher.services.impl;

import by.bsuir.dc.publisher.dal.MessageDao;
import by.bsuir.dc.publisher.services.exceptions.ApiException;
import by.bsuir.dc.publisher.services.exceptions.GeneralSubCode;
import by.bsuir.dc.publisher.services.impl.mappers.MessageMapper;
import by.bsuir.dc.publisher.entities.Message;
import by.bsuir.dc.publisher.entities.dtos.request.MessageRequestTo;
import by.bsuir.dc.publisher.entities.dtos.response.MessageResponseTo;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestClient;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

//@Service
@RequiredArgsConstructor
public class MessageServiceImpl {

    private final String discussionUri;

    private final MessageDao dao;

    private final RestClient restClient;

    private final MessageMapper mapper;

    private String getCountry(String lang) {
        int commaIndex = lang.indexOf(',');
        String country = lang;

        if (commaIndex != -1) {
            country = lang.substring(0, commaIndex);
        }

        return country;
    }

    public MessageResponseTo create(MessageRequestTo requestTo, String lang)
            throws ApiException {

        String country = getCountry(lang);

        Message model = mapper.requestToModel(requestTo);
        model.setCountry(country);

        ResponseEntity<Message> result = restClient
                .post()
                .uri(discussionUri)
                .body(model)
                .retrieve()
                .onStatus(
                    HttpStatusCode::is5xxServerError,
                    (request, response) -> {
                        throw new ApiException(
                                response.getStatusCode().value(),
                                1,//change
                                "Some error message"//change here
                        );
                    }
                )
                .toEntity(Message.class);

        Message savingRes = result.getBody();
        return mapper.modelToResponse(savingRes);
    }

    public List<MessageResponseTo> getAll() {
        Iterable<Message> models = dao.findAll();
        return StreamSupport
                .stream(models.spliterator(), false)
                .map(mapper::modelToResponse)
                .collect(Collectors.toList());
    }

    public MessageResponseTo get(Long id) throws ApiException {
        Optional<Message> author = dao.findById(id);

        if (author.isEmpty()) {
            throw new ApiException(
                    HttpStatus.NOT_FOUND.value(),
                    GeneralSubCode.WRONG_ID.getSubCode(),
                    GeneralSubCode.WRONG_ID.getMessage()
            );
        }

        return mapper.modelToResponse(author.get());
    }

    public MessageResponseTo update(MessageRequestTo requestTo) throws ApiException {
        Message model = mapper.requestToModel(requestTo);

        Message updateRes = dao.update(model);
        //if (updateRes == null) {
        //    throw new ApiException(
        //            HttpStatus.NOT_FOUND.value(),
        //            GeneralSubCode.WRONG_ID.getSubCode(),
        //            GeneralSubCode.WRONG_ID.getMessage()
        //    );
        //}

        return mapper.modelToResponse(updateRes);
    }

    public void delete(Long id) throws ApiException {
        if (!dao.existsById(id)) {
            throw new ApiException(
                    HttpStatus.NOT_FOUND.value(),
                    GeneralSubCode.WRONG_ID.getSubCode(),
                    GeneralSubCode.WRONG_ID.getMessage()
            );
        }

        dao.deleteById(id);
    }

}

File: 151004/Burbouski/appLab2/src/main/kotlin/com/github/hummel/dc/lab2/service/impl/MessageServiceImpl.kt
package com.github.hummel.dc.lab2.service.impl

import com.github.hummel.dc.lab2.bean.Message
import com.github.hummel.dc.lab2.dto.request.MessageRequestTo
import com.github.hummel.dc.lab2.dto.request.MessageRequestToId
import com.github.hummel.dc.lab2.dto.response.MessageResponseTo
import com.github.hummel.dc.lab2.service.MessageService
import com.github.hummel.dc.lab2.util.BaseRepository

class MessageServiceImpl(
	private val messageRepository: BaseRepository<Message, Long>
) : MessageService {
	override fun getAll(): List<MessageResponseTo> {
		val result = messageRepository.data.map { it.second }

		return result.map { it.toResponse() }
	}

	override fun create(messageRequestTo: MessageRequestTo?): MessageResponseTo? {
		val id = if (messageRepository.data.isEmpty()) {
			-1
		} else {
			messageRepository.getLastItem()?.id ?: return null
		} + 1

		val bean = messageRequestTo?.toBean(id) ?: return null
		val result = messageRepository.addItem(bean.id, bean)

		return result?.toResponse()
	}

	override fun deleteById(id: Long): Boolean = messageRepository.removeItem(id)

	override fun getById(id: Long): MessageResponseTo? {
		val result = messageRepository.getItemById(id)?.second

		return result?.toResponse()
	}

	override fun update(messageRequestToId: MessageRequestToId?): MessageResponseTo? {
		val bean = messageRequestToId?.toBean() ?: return null
		val result = messageRepository.addItem(bean.id, bean)

		return result?.toResponse()
	}
}